becquerel/core/rebin.py

```diff
@@ -0,0 +1,433 @@
+        array of the input bin edges of shape
+      out_edges: np.1darray
+        array of the output bin edges
+      slopes: TODO unused, just for keeping number of arguments the same as
+              _rebin_interpolation()
+
+    Returns:
+      np.2darray, shape (num_spectra, num_channels_in)
+      The rebinned spectra
+    """
+    # Init output
+    out_spectra = np.zeros((in_spectra.shape[0], out_edges.shape[0] - 1),
+                           np.int64)
+    for i in np.arange(in_spectra.shape[0]):
+        out_spectra[i, :] = _rebin_listmode(in_spectra[i, :], in_edges[i, :],
+                                            out_edges, slopes[i, :])
+    return out_spectra
+
+
+def rebin(in_spectra, in_edges, out_edges, method="interpolation",
+          slopes=None, zero_pad_warnings=True):
+    """
+    Spectra rebining via deterministic or stochastic methods.
+
+    Args:
+        in_spectrum (np.ndarray): an array of input spectrum counts (1D or 2D)
+            [num_spectra, num_channels_in] or [num_channels_in]
+        in_edges (np.ndarray): an array of the input bin edges (1D or 2D)
+            [num_spectra, num_channels_in + 1] or [num_channels_in + 1]
+        out_edges (np.ndarray): an array of the output bin edges
+            [num_channels_out]
+        method (str): rebinning method
+            "interpolation"
+                Deterministic interpolation
+            "listmode"
+                Stochastic rebinning via conversion to listmode of energies
+        slopes (np.ndarray|None): (optional) an array of input bin slopes for
+            quadratic interpolation (1D or 2D)
+            (only applies for "interpolation" method)
+            [num_spectra, num_channels_in + 1] or [num_channels_in + 1]
+        zero_pad_warnings (boolean): warn when edge overlap results in
+            appending empty bins
+
+    Raises:
+        AssertionError: for bad input arguments
+
+    Returns:
+        The rebinned spectrum/a
+    """
+    method = method.lower()
+    if method == "listmode":
+        try:
+            in_spectra = np.asarray(in_spectra).astype(
+                np.int64, casting="safe", copy=False)
+        except TypeError:
+            raise ValueError(
+                "in_spectrum can only contain ints for method listmode")
+    else:
+        in_spectra = np.asarray(in_spectra, np.float)
+    in_edges = np.asarray(in_edges, np.float)
+    out_edges = np.asarray(out_edges, np.float)
+    # Check inputs
+    # TODO check that in_spectrum are all >= 0
+    _check_ndim(in_spectra, {1, 2}, 'in_spectra')
+    # broadcast in_edges out to the dimensions of in_spectra
+    # specifically for the case 1D -> 2D
+    if (in_spectra.ndim == 2) and (in_edges.ndim == 1):
+        # copy req'd: the readonly array doesn't work w/ numba
+        in_edges = np.copy(
+            np.broadcast_to(
+                in_edges,
+                (in_spectra.shape[0], in_edges.shape[-1])))
+    _check_ndim(out_edges, 1, 'out_edges')
```

Comment:
@jccurtis Well done putting this in, very elegant. Is there a particular reason you added a flag for `zero_pad`, wouldn't it be better to add a `no_checks` flag instead which removes all the checks in one go (for performance if necessary).

Reply:
Flag suppresses warnings, other tests raise errors which are not disabled. That makes sense.